src/app/content/highlights/components/CardWrapper.tsx

```diff
@@ -11,20 +15,109 @@ interface Props {
   className?: string;
 }
 
+const displayAllCardsMediaQuery = `(max-width: ${mediaQueryBreakToStopDisplaingAllCards})`;
+
 // tslint:disable-next-line:variable-name
-const Wrapper = ({highlights, className, container, highlighter}: Props) => highlights.length
-  ? <div className={className}>
+const Wrapper = ({highlights, className, container, highlighter}: Props) => {
+  const displayAllCards = !assertWindow().matchMedia(displayAllCardsMediaQuery).matches;
+
+  const element = React.useRef<HTMLElement>(null);
+  const [cardsPositions, setCardsPositions] = React.useState<Map<string, number>>(new Map());
+  const [cardsHeights, setCardsHeights] = React.useState<Map<string, number>>(new Map());
+
+  const onHeightChange = (id: string, ref: React.RefObject<HTMLElement>) => {
+    if (!displayAllCards) { return; }
+
+    const height = ref.current && ref.current.offsetHeight;
+    if (cardsHeights.get(id) !== height) {
+      setCardsHeights((data) => new Map(data.set(id, height === null ? 0 : height)));
+    }
+  };
+
+  const onFocus = (id: string) => {
+    if (!displayAllCards) { return; }
+
+    const highlight = highlights.find((search) => search.id === id);
+    const position = cardsPositions.get(id);
+    if (typeof position !== 'number' || !highlight) { return; }
+
+    const topOffset = assertDefined(
+      getHighlightTopOffset(container, highlight),
+      `Couldn't get top offset for highlights`
+    );
+
+    if (position > topOffset) {
+      element.current!.style.top = `-${position - topOffset}px`;
+    }
+
+    scrollIntoView(highlight.elements[0] as HTMLElement);
+  };
+
+  const onBlur = () => {
+    element.current!.style.top = '0';
+  };
+
+  const updatePositions = React.useCallback(() => {
+    const newPositions: Map<string, number> = new Map();
+
+    let lastVisibleCardPosition: number = 0;
+    let lastVisibleCardHeight: number = 0;
+
+    for (const [index, highlight] of highlights.entries()) {
+      const topOffset = assertDefined(
+        getHighlightTopOffset(container, highlight),
+        `Couldn't get top offset for highlights`
+      );
+
+      let stackedTopOffset = lastVisibleCardPosition;
+
+      if (topOffset < (lastVisibleCardPosition + lastVisibleCardHeight + remsToPx(cardMarginBottom))) {
+        stackedTopOffset = stackedTopOffset
+          + lastVisibleCardHeight
+          + (index > 0 ? remsToPx(cardMarginBottom) : 0);
+      } else {
+        stackedTopOffset = topOffset;
+      }
+
+      if (cardsHeights.get(highlight.id)) {
```

Comment:
It will be `undefined` for the first render, before cards will fire `onHeightChange` which should happen after they render

Reply:
would it be more correct to noop earlier in the logic for that case then?

Reply:
Currently, it will calculate positions for cards depends on the default heigh value and then it will update them only if they are different so I think it is okay to do not noop this